clang/test/Sema/builtin-redecl.cpp

```diff
@@ -14,13 +14,18 @@ void __builtin_va_copy(double d);
 // expected-error@+2 {{cannot redeclare builtin function '__builtin_va_end'}}
 // expected-note@+1 {{'__builtin_va_end' is a builtin with type}}
 void __builtin_va_end(__builtin_va_list);
-// RUN: %clang_cc1 %s -fsyntax-only -verify 
-// RUN: %clang_cc1 %s -fsyntax-only -verify -x c
 
 void __va_start(__builtin_va_list*, ...);
 
+          void *__builtin_assume_aligned(const void *, size_t, ...);
 #ifdef __cplusplus
-void *__builtin_assume_aligned(const void *, size_t, ...) noexcept;
-#else
-void *__builtin_assume_aligned(const void *, size_t, ...);
+constexpr void *__builtin_assume_aligned(const void *, size_t, ...);
+          void *__builtin_assume_aligned(const void *, size_t, ...) noexcept;
+constexpr void *__builtin_assume_aligned(const void *, size_t, ...) noexcept;
+          void *__builtin_assume_aligned(const void *, size_t, ...) throw();
+constexpr void *__builtin_assume_aligned(const void *, size_t, ...) throw();
+
```

Comment:
What is the motivation to allow that?

I kind of understand letting users defining builtin (actually, I am not sure I do). but once the builtin is redefined once, I think should be some sort of consistency with the language rules.
Here if you were to replace `__builtin_assume_aligned` with `f`, it would be ill-formed, for god reason.

Reply:
For backwards compatibility (and compatibility with the same definitions in C or with GCC), to allow declarations with or without the `constexpr`. I suspect in any real code, it will always be declared without the `constexpr`.

The other solutions is to keep them all as `ConstexprSpecKind::Unspecified` like we currently do, but this is asymmetric with `consteval` builtins which will be marked `ConstexprSpecKind::Consteval`, and leads to the builtin versions of `std::move`/`std::forward`/constexpr cmath functions to have differing `getConstexprKind()` from their non-builtin counterparts

Reply:
Does anyone know why we allow builtins to be redeclared at all? CC @zygoloid @rjmccall 

I wonder just how much code we would break by not aiming for backwards compatibility here and requiring users to update their declarations on the assumption that declaring these at all is undefined behavior (using a reserved name) and so it's up to the user to match the compiler's behavior rather than make the compiler handle this with weird non-conforming language extension behavior.

I'm not looking to punish users in this case, but the workaround to keep things backwards compatible is pretty magical and awkward.

Reply:
I don't think there are particularly good non-historical reasons why we allow non-`LIBBUILTIN` builtins to be redeclared -- I think it's allowed primarily for compatibility with GCC and existing code, and it's not clear to me why GCC allows it. There were some examples of code redeclaring builtins in [this prior review thread](https://reviews.llvm.org/D45383).

Reply:
@erichkeane

Reply:
From reading through that (and trying to search my memory): A number of our `LIBBUILTIN` builtins are just replacements for things that would otherwise be implemented in the  library.  

So we need to allow them to be redeclared because they might be implemented in the library.

However, I would be unopposed to a patch that makes declaring a `__builtin` spelled as  reserved identifier ill-formed.

Reply:
That seems like that should go in a separate patch. For now, I'll change this one so `ConstexprSpecKind::Constexpr` isn't added to these builtins as before